### ISSUE TITLE
"All Accounts" filters' section, needed

### ISSUE THREAD

**Issue by xlovinglyx** [RealRaven2000/FiltaQuilla#264]
Instead of playing with filters files & copying them around when another email account is created,
I'd like filtaquilla to apply the new "All Accounts" filters i create to all present & future acccounts
"Available".... .

Tools/
Message Filters/
Filters for/

"All Accounts"

**Comment by RealRaven2000** [comment]
Thanks for your suyggestion. However, as I said in email, it's probably a lot better for Thunderbird itself to implement this - they are the ones that store all filters and they execute them through a filter service (which is part of the compiled code running in C++) - this part of the code cannot be altered by any Add-on because there are only certain entry points that are open to scripting (which is done with JavaScript through APIs and the older scriptable XUL interfaces).

 I think the very best place that discussed this requirement is here:

https://bugzilla.mozilla.org/show_bug.cgi?id=34973

it would be fantastic if you could add a comment there, as old bugs are more likely to be looked at when they get a bit of traffic. By the way none of the features done by FiltaQuilla could be done using the official web extension APIs, so any modern Add-on wouldn't even be able to implement such features as regular expressions or running external programs. FiltaQuilla uses the  "grandfathered" way of accessing all the top layer scripting which forces it into "experimental" status, including the big scary warning that's shown when it is installed. This is another indicator that this would be much better implemented by Thunderbird itself.

I will ask about this feature in tonight's add-on developer Zoom call and will mention it to John (Thunderbird employee) to ask whether he sees a chance that Thunderbird will make resources available to implement it. I hear they are thinking of making filters async which is good news because only then the APIs can start to expose them to Add-ons. Not sure if this will land in Thunderbird 128 or will take until next year.

If you want an alternative mail client based on the Mozilla code base that implements this feature, please take a look at Postbox. They implement this feature by copying the filter to all accounts whenever it changes.

Unfortunately they stopped supporting 3rd party Add-ons because it created too many support issues (wasting their time) so  that they couldn't concentrate on improving their product. But maybe they implemented Regular Expressions in their filters, although I am not sure they love answering questions on that (?). Regular Expressions are hard and for that reason most developers stay well from them.

**Comment by xlovinglyx** [comment]
1. mozilla staff ? it will take a year up to a decade (i tell u that by experience), useless.
2. postbox method ? it enlarges the stupid global features in the memory (infantile method)
3. "FiltaQuilla uses the "grandfathered" way of accessing all the top layer scripting which forces it into "experimental" status, including the big scary warning that's shown when it is installed. This is another indicator that this would be much better implemented by Thunderbird itself."
It doesn't mean you or friends helping you with filtaquilla can't.. & this would be the most profitable way of having this feature
available sooner.

**Comment by RealRaven2000** [comment]
>   1. mozilla staff ? it will take a year up to a decade (i tell u that by experience), useless.

well, at least they are paid for their work if they choose to work on filters. I know that they will tackle making filtering async that means it will become more accessible to (web) extensions so I there will be more competition, other developers adding functionality. It would be good to find somebody in Thunderbird staff who is specialized in the filtering aspects, I have many things on my personal wish list. Unique IDs and custom attributes would be at the very top.

> 
>     2. postbox method ? it enlarges the stupid global features in the memory (infantile method)

I haven't looked at the Postbox codebase for a few years, so I don't know.

> 
>     3. "FiltaQuilla ... this would be much better implemented by Thunderbird itself."

>        It doesn't mean you or friends helping you with filtaquilla can't.. & this would be the most profitable way of having this feature available sooner.

I asked John, he is not directly involved with filtering but he may touch the field after May. 

Currently there is a code sprint going on as they are freezing all new feature development at the end of the month to prepare for the release of 128 this summer. They certainly won't do very quick hacks that add functionality and may prove as unstable or make the code even more convoluted. For risky quick changes Betterbird would be a good alternative.

**Comment by xlovinglyx** [comment]
keep me in touch if John can do something about it. thanks.

**Comment by xlovinglyx** [comment]
i'll try to see if there's an email client who has this involved as an alternative to tb too, cause tb is a very bad architecture
that involves memory issues & doesn't exploit the hd & vm totally & conveniently. even with a long subject list, when i try
to edit it, it freezes FOREVER.. & i experienced that frreezing issue with palemoon too aka firefox in many situations through
various updates.

**Comment by RealRaven2000** [comment]
> keep me in touch if John can do something about it. thanks.

Yeah he said he can't (at least before the code freeze) and he is not specialized in this area (filter service). But I will keep mentioning it, after the code sprint. Not point in stressing people out when they don't have extra free time and are scrambling to finish stuff for QA. They are slowly removing the old code from Netscape, I am not completely without hope!

They are planning to put the code base on Github and move away from the horrible build system they have now,  then I will able to look into submitting patches to Thunderbird itself. It's just  pointless trying to write patches when building takes ages and you can't really debug it. (I had done some patches to the top layer in Tb, but all strictly JavaScript, as their old C++code is horrible)

**Comment by xlovinglyx** [comment]
evolution has regex in filters but the ui is shit. i hope this fix won't take forever.

**Comment by xlovinglyx** [comment]
& ps: the bugzilla thread is a 24 year old thread, just to show u how much tos they give about it.

**Comment by xlovinglyx** [comment]
There's also a plugin for that in thebat x64 too, but i can't get to have thebat offer an all in one inbox..

**Comment by RealRaven2000** [comment]
> There's also plugin for that in thebat x64 too, but i can't get to have thebat offer an all in one inbox..

interesting, Although I don't think  the bat is based on Mozilla / Netscape code? by the way, I believe the current dedicated Thunderbird team / in its current form is only 3 years old and they have a 1.4 Gbyte code base (comm central) plus have to react to all the changes handed down from the Firefox code base (mozilla central) as well.

I think they are improving recently and hopefully Thunderbird will be stable within the next 24 months. There is also a mobile version (Thunderbird K9) which they are working on.

**Comment by xlovinglyx** [comment]
I have no idea when this issue will be fixed for tb & i've shown u the attitude towards the issue the mozilla staff had..
24 years of waiting & still no step forward .. . as you can see: "common filters", will filter all emails without creating lots of filter files per account. & it's memory friendly when you try to edit long lists, I've found my beast & waiting ages for a specific update is not my cup of tea.. look at this beast:

![image](https://github.com/RealRaven2000/FiltaQuilla/assets/53103638/dbdb2931-9e65-41a7-9cda-372aeda10ae1)

**Comment by RealRaven2000** [comment]
You can use Betterbird for building more complex structures in filters, like this:

![image](https://github.com/RealRaven2000/FiltaQuilla/assets/6314674/7acd09e8-0ada-460a-ac6d-7393f6e51ea9)

They used the existing data structure for this so that's why it looks a little weird; but you can build combined filters with this.

I helped them a little with UI suggestions. The code was hacky though, so Thunderbird will have to rewrite this in a more organized fashion. The only problem here is that it may break my "merge" function on my Add-on quickFilters.

**Comment by xlovinglyx** [comment]
i don't want betterbird if there's no regex.

**Comment by RealRaven2000** [comment]
> i don't want betterbird if there's no regex.

you can run all Add-ons on Betterbird - this includes FiltaQuilla with its regular expressions.

It's very similar to Thunderbird they just cherry pick some of the more urgent (and sometimes old) bugs and hack them quickly to give the functionality to the user, without too much regard for future code maintainability. Thunderbird goes a mopre managed and "safe" route because they don't want to pile more "shit" on the code base, so when they implement a feature it is redesigned from the ground to avoid regressions.

**Comment by xlovinglyx** [comment]
does it have an all accounts filter section ? if no, i'm staying on voyager.

**Comment by RealRaven2000** [comment]
> does it have an all accounts filter section ? if no, i'm staying on voyager.

no, that's Postbox. But no Add-ons in that. good look with Voyager!!

**Comment by TonyGravagno** [comment]
Axel, your Thanks note on 9 May gave me an idea about this ticket. 
It's not necessary to add "All Accounts" to the filter selection

In **quickFilters** youalready have the feature to backup and restore filters. What do you think about this?

- Allow user to select filters and a destination account.
- Serialize desired filters to JSON ( = backup without save to disk, or save to disk if required )
- Change the `accountName`, `rootFolderURL`, and `actionList>targetFolderUri` nodes as required.
- Append a unique suffix to all `filterName` values to ensure filters are not over-written.
- Import them like a restore.

Thoughts?
